data-prepper-core/src/main/java/com/amazon/dataprepper/parser/PipelineParser.java

```diff
@@ -54,7 +54,9 @@ public PipelineParser(final String configurationFileLocation) {
                     new TypeReference<Map<String, PipelineConfiguration>>() {
                     });
             final List<String> allPipelineNames = PipelineConfigurationValidator.validateAndGetPipelineNames(pipelineConfigurationMap);
-            final Map<String, Pipeline> pipelineMap = new HashMap<>();
+
+            // LinkedHashMap to preserve insertion order
+            final Map<String, Pipeline> pipelineMap = new LinkedHashMap<>();
```

Comment:
QUES: Is the insertion order guaranteed to be topological? It seems a little fragile but have not checked the context carefully.

Reply:
It is, but only because we're calling [validateAndGetPipelineNames](https://github.com/opendistro-for-elasticsearch/data-prepper/blob/e2b858c0f15928bf28857d44d8b4f567aa973f49/data-prepper-core/src/main/java/com/amazon/dataprepper/parser/PipelineConfigurationValidator.java#L22-L31) which does the sorting for us.

I agree it's a bit fragile - maybe a better approach would have the `shutdown` logic perform the sorting just in case the "validate" method ever stops sorting. I used what already existed here though as a shortcut.

Reply:
I do not see a use case where we will disable the validate, why are we calling it fragile ? I agree that when we implemented it, we implemented as a bonus but not as a requirement. 
Although it ensures that pipelines are always in topological order, there is a case where if we have orphan pipelines - each pipeline will be in topological order but pipeline ids can be interleaved, which I think will not be an issue - if we are shutting down everything.

Example: [_very rare scenario but possible depending on the order of pipelines defined in yml file_]
P1 -> P2 -> P3 -> P4
Pa -> Pb -> Pc
The order can be like  -> P1, P2, Pa, P3, Pb, P4, Pc.

We also had a commented method to validate orphans and fail but I think we removed it.

Reply:
Ah the logic is fine - I'm glad `validate` was implemented, it saved me a lot of effort :) Also interleaved is fine too, just as long as each individual pipeline chain is shutdown in-order from upstream to downstream.

We're calling my implementation here "fragile" because currently there exists a slight temporal coupling between `validateAndGetPipelineNames` and `shutdown`. In this case, `shutdown` _relies_ on some other operation to perform the topological sort and persist those values in the hashmap. `shutdown` has a soft requirement that `validate` be called prior. It would be more robust if `shutdown` instead performed whichever actions it needed (sorting) as part of its routine.

I agree it's unlikely, but if for whatever reason someone comes along and does a refactor which causes the pipelineMap to no longer be sorted, `shutdown`'s behavior will be impacted in a non-obvious way.